cleverhans/future/tf2/attacks/fast_gradient_method.py

```diff
@@ -0,0 +1,135 @@
+  :param clip_max: (optional) float. Maximum float value for adversarial example components.
+  :param y: (optional) Tensor with true labels. If targeted is true, then provide the
+            target label. Otherwise, only provide this parameter if you'd like to use true
+            labels when crafting adversarial samples. Otherwise, model predictions are used
+            as labels to avoid the "label leaking" effect (explained in this paper:
+            https://arxiv.org/abs/1611.01236). Default is None.
+  :param targeted: (optional) bool. Is the attack targeted or untargeted?
+            Untargeted, the default, will try to make the label incorrect.
+            Targeted will instead try to move in the direction of being more like y.
+  :param sanity_checks: bool, if True, include asserts (Turn them off to use less runtime /
+            memory or for unit tests that intentionally pass strange input)
+  :return: a tensor for the adversarial example
+  """
+  if ord not in [np.inf, 1, 2]:
+    raise ValueError("Norm order must be either np.inf, 1, or 2.")
+
+  asserts = []
+
+  # If a data range was specified, check that the input was in that range
+  if clip_min is not None:
+    asserts.append(tf.math.greater_equal(x, clip_min))
+
+  if clip_max is not None:
+    asserts.append(tf.math.less_equal(x, clip_max))
+
+  if y is None:
+    # Using model predictions as ground truth to avoid label leaking
+    y = tf.argmax(model_fn(x), 1)
+
+  grad = train_step(model_fn, x, y, targeted)
+  optimal_perturbation = optimize_linear(grad, eps, ord)
+  # Add perturbation to original example to obtain adversarial example
+  adv_x = x + optimal_perturbation
+
+  # If clipping is needed, reset all values outside of [clip_min, clip_max]
+  if (clip_min is not None) or (clip_max is not None):
+    # We don't currently support one-sided clipping
+    assert clip_min is not None and clip_max is not None
+    adv_x = tf.clip_by_value(adv_x, clip_min, clip_max)
+
+  if sanity_checks:
+    assert np.all(asserts)
+  return adv_x
+
+
+@tf.function
+def train_step(model_fn, x, y, targeted):
+  """
+  Calculates the gradients with respect to the input x. Wrapped inside
+  of @tf.function to speed up computation.
+  :param model_fn: a callable that takes an input tensor and returns the model logits.
+  :param x: input tensor.
+  :param y: (optional) Tensor with true labels. If targeted is true, then provide the
+            target label. Otherwise, only provide this parameter if you'd like to use true
+            labels when crafting adversarial samples. Otherwise, model predictions are used
+            as labels to avoid the "label leaking" effect (explained in this paper:
+            https://arxiv.org/abs/1611.01236). Default is None.
+  :param targeted: (optional) bool. Is the attack targeted or untargeted?
+            Untargeted, the default, will try to make the label incorrect.
+  """
+  loss_fn = tf.losses.SparseCategoricalCrossentropy(from_logits=True)
```

Comment:
For stateless objects, we prefer to use functional APIs typically from `tf.nn.*` such as `tf.nn.softmax_cross_entropy_with_logits`.